### PR TITLE
Save removed artifacts list to json file

### PR DESCRIPTION
**Pull request by roytev**
Adds the ability to save the full output of artifacts_to_remove per policy to json file.
i want to save the files for report generation and for auditing.

### CODE REVIEW

File: artifactory_cleanup/artifactorycleanup.py [hunk 1]
```diff
@@ -75,6 +82,9 @@ def cleanup(self, block_ctx_mgr, test_ctx_mgr) -> Iterator[CleanupSummary]:
                 print("Summary size not defined")
                 yield None
             print()
+            if self.save_removed_artifacts_list:
+                with open(os.path.join(self.save_removed_artifacts_path, f"{policy.name}.json"), 'w') as file:
```
Comment (allburov) [inline artifactory_cleanup/artifactorycleanup.py:86]: `policy.name` can contain unsupported characters by OS, we should escape it first

[the remark this answers is not in the document]
Comment (allburov) [reply artifactory_cleanup/artifactorycleanup.py:86]: or `--output-format` as we already do

**Comment by allburov** [comment]
I see, the current `--output` does it at the end with the summary.

We should rename the current one to something like `--output-removed` with `--output-removed-format` may be (which is only JSON for now)

**Comment by allburov** [comment]
Also in the current approach if network stop working in the middle - you'll lose all previously removed artifacts for the policy, no file will be created.
We handle it by either add different format where we could add it line by line right when we removed it (like CSV) or adding track really removed artifacts and write them down at the end

Or just ignore that case, but we should document it :)

File: artifactory_cleanup/artifactorycleanup.py
Comment (allburov) [inline artifactory_cleanup/artifactorycleanup.py:1]: BTW, could you write a few lines in README.md when we agreed on other changes

File: artifactory_cleanup/cli.py [hunk 2]
```diff
@@ -184,9 +195,9 @@ def main(self):
                     "name": summary.policy_name,
                     "file_count": summary.artifacts_removed,
                     "size": summary.artifacts_size,
+                    **({"removed_artifacts_list": summary.removed_artifacts_list} if summary.removed_artifacts_list is not None else {})
```
Comment (allburov) [inline artifactory_cleanup/cli.py:198]: Obviosly it's removed and list, lets simplify the key to `artifacts`

Comment (allburov) [inline artifactory_cleanup/cli.py:198]: > **({"removed_artifacts_list": summary.removed_artifacts_list} if summary.removed_artifacts_list is not None else {})

I don't like that constrcution - too much actions in a single line

Please make it simplier by using additional variables. 
We can always add `artifacts` key, but just set it to `null` if saving not required

File: README.md [hunk 3]
```diff
@@ -151,6 +151,9 @@ artifactory-cleanup --output=myfile.txt
 
 # Save the summary in a Json file
 artifactory-cleanup --output=myfile.txt --output-format=json
+
+# Save the summary and removed artifacts list in a Json file
+artifactory-cleanup --output=myfile.txt --output-format=json --save-removed-artifacts
```
Comment (allburov) [inline README.md:156]: The last argument looks like it doesn't affect `--output`, but it does.

We should name it in a more generic way, like 
```suggestion
artifactory-cleanup --output=myfile.txt --output-format=json --output-artifacts 
```

File: artifactory_cleanup/artifactorycleanup.py [hunk 4]
```diff
@@ -69,6 +75,7 @@ def cleanup(self, block_ctx_mgr, test_ctx_mgr) -> Iterator[CleanupSummary]:
                     policy_name=policy.name,
                     artifacts_size=artifacts_size,
                     artifacts_removed=len(artifacts_to_remove),
+                    removed_artifacts_list=artifacts_to_remove if self.save_removed_artifacts_list else None,
```
Comment (allburov) [inline artifactory_cleanup/artifactorycleanup.py:78]: Let's use the boolean, why do we need to convert it to None, don't get it

Comment (allburov) [reply artifactory_cleanup/artifactorycleanup.py:78]: The default is False, so it's fine for us

File: artifactory_cleanup/cli.py (hunk 2 above)
Comment (allburov) [reply artifactory_cleanup/cli.py:198]: at the end it should be like 
```suggestion
                    "artifacts": summary.removed_artifacts
```

**Review by allburov — COMMENTED** [review]
Hi! Let's go one more time and adjust the code a bit if you don't mind :)

**Comment by roytev** [comment]
Hi @allburov Thanks again! ive fixed according to your review :)

File: artifactory_cleanup/artifactorycleanup.py [hunk 5]
```diff
@@ -65,12 +71,14 @@ def cleanup(self, block_ctx_mgr, test_ctx_mgr) -> Iterator[CleanupSummary]:
...
-                yield CleanupSummary(
+                summary = CleanupSummary(
                     policy_name=policy.name,
                     artifacts_size=artifacts_size,
                     artifacts_removed=len(artifacts_to_remove),
                 )
-
+                if self.save_removed_artifacts_list:
+                    summary.removed_artifacts_list = artifacts_to_remove
+                yield summary
```
Comment (allburov) [inline artifactory_cleanup/artifactorycleanup.py:81]: LEt's always transfer that data, so Cleanup even doesn't have to worry about `self.save_removed_artifacts_list`
It's inmemory and cheep data

File: artifactory_cleanup/cli.py [hunk 6]
```diff
@@ -179,14 +189,14 @@ def main(self):
...
-                }
-            )
-
+            policy = {
+                "name": summary.policy_name,
+                "file_count": summary.artifacts_removed,
+                "size": summary.artifacts_size
+            }
+            if summary.removed_artifacts_list is not None:
+                policy["artifacts_list"] = summary.removed_artifacts_list
```
Comment (allburov) [inline artifactory_cleanup/cli.py:198]: no needs to specify the type 
```suggestion
                policy["artifacts"] = summary.removed_artifacts_list
```

File: artifactory_cleanup/artifactorycleanup.py [hunk 7]
```diff
@@ -13,6 +15,7 @@ class CleanupSummary:
     policy_name: str
     artifacts_removed: int
     artifacts_size: int
+    removed_artifacts_list: Optional[dict] = None
```
Comment (allburov) [inline artifactory_cleanup/artifactorycleanup.py:18]:
```suggestion
    removed_artifacts Optional[dict] = None
```

Is it a `dict?` :thinking:

File: artifactory_cleanup/cli.py [hunk 8]
```diff
@@ -179,14 +189,14 @@ def main(self):
...
-                    "size": summary.artifacts_size,
-                }
-            )
-
+            policy = {
+                "name": summary.policy_name,
+                "file_count": summary.artifacts_removed,
+                "size": summary.artifacts_size
+            }
+            if summary.removed_artifacts_list is not None:
```
Comment (allburov) [inline artifactory_cleanup/cli.py:197]:
```suggestion
            if self._output_artifacts:
```
Let's use this one

**Comment by allburov** [comment]
Merged in https://github.com/devopshq/artifactory-cleanup/pull/145

**Pull request closed without merging**